bookstore/clone.py

```diff
@@ -40,6 +37,68 @@ class BookstoreCloneHandler(IPythonHandler):
     `Jupyter Notebook reference on Custom Handlers <https://jupyter-notebook.readthedocs.io/en/stable/extending/handlers.html#registering-custom-handlers>`_
     """
 
+    def initialize(self):
+        """Helper to retrieve bookstore setting for the session."""
+        self.bookstore_settings = BookstoreSettings(config=self.config)
+
+    @web.authenticated
+    async def get(self):
+        """GET /bookstore/clone?s3_bucket=<your_s3_bucket>&s3_key=<your_s3_key>
+
+        Renders an options page that will allow you to clone a notebook
+        from a specific bucket via the Bookstore cloning API.
+        """
+        s3_bucket = self.get_argument("s3_bucket")
+        if s3_bucket == '' or s3_bucket == "/":
+            raise web.HTTPError(400, "Must have a bucket to clone from")
+
+        # s3_paths module has an s3_key function; s3_object_key avoids confusion
+        s3_object_key = self.get_argument("s3_key")
+        if s3_object_key == '' or s3_object_key == '/':
+            raise web.HTTPError(400, "Must have a key to clone from")
+
+        self.log.info("Setting up cloning landing page from %s", s3_object_key)
+
+        template_params = self.construct_template_params(s3_bucket, s3_object_key)
+        self.set_header('Content-Type', 'text/html')
+        self.write(self.render_template('clone.html', **template_params))
+
+    def construct_template_params(self, s3_bucket, s3_object_key):
+        """Helper that takes valid S3 parameters and populates UI template"""
+        base_uri = f"{self.request.protocol}://{self.request.host}"
+        clone_api_url = url_path_join(base_uri, self.base_url, "/api/bookstore/cloned")
```

Comment:
Should this end with clone?

Reply:
Indeed!

Reply:
It also should have began with bookstore! yay for tests actually failing!